pkg/adapter/adapter.go

```diff
@@ -31,17 +36,141 @@ type envConfig struct {
+	SinkURL                string `envconfig:"K_SINK" required:"false"`
 }
 
-// NewEnvConfig creates an empty configuration
+type mongoDbAdapter struct {
+	namespace       string
+	ce              cloudevents.Client
+	source          string
+	database        string
+	collection      string
+	credentialspath string
+	sinkurl         string
+	logger          *zap.SugaredLogger
+}
+
+// dataSource interface to interact with either a mongo.database or a mongo.collection.
+type dataSource interface {
+	Watch(ctx context.Context, pipeline interface{}, opts ...*options.ChangeStreamOptions) (*mongo.ChangeStream, error)
+}
+
+// NewEnvConfig creates an empty environement variables configuration.
 func NewEnvConfig() adapter.EnvConfigAccessor {
 	return &envConfig{}
 }
 
 // NewAdapter creates an adapter to convert incoming MongoDb changes events to CloudEvents and
-// then sends them to the specified Sink
+// then sends them to the specified Sink.
 func NewAdapter(ctx context.Context, processed adapter.EnvConfigAccessor, ceClient cloudevents.Client) adapter.Adapter {
-	fmt.Printf("Hey from mongodb adapter")
-	time.Sleep(600 * time.Second)
+	logger := logging.FromContext(ctx)
+	env := processed.(*envConfig)
+
+	return &mongoDbAdapter{
+		namespace:       env.Namespace,
+		ce:              ceClient,
+		source:          env.EventSource,
+		database:        env.Database,
+		collection:      env.Collection,
+		credentialspath: env.MongoDbCredentialsPath,
+		sinkurl:         env.SinkURL,
+		logger:          logger,
+	}
+}
+
+// Start connects to the database and creates the watch stream that will watch for dataSource changes.
+func (a *mongoDbAdapter) Start(ctx context.Context) error {
+	// Read the Credentials.
+	rawURI, err := ioutil.ReadFile(a.credentialspath + "/URI")
+	if err != nil {
+		return fmt.Errorf("Unable to get MongoDb URI field: secretPath %s : %w", a.credentialspath+"/URI", err)
+	}
+	URI := string(rawURI)
+	if URI == "" {
+		return fmt.Errorf("MongoDb URI field is an empty string: URI %s : %w", URI, err)
+	}
+
+	// Create new Client.
+	client, err := mongo.NewClient(options.Client().ApplyURI(URI))
+	if err != nil {
+		return fmt.Errorf("Error creating mongo client: %w", err)
+	}
+	// Get dataSource
+	dataSource := getDataSource(client, a.database, a.collection)
+
+	// Connect to Client.
+	err = client.Connect(ctx)
+	if err != nil {
+		return fmt.Errorf("Error connecting to database: %s", err)
+	}
+	defer client.Disconnect(ctx)
+
+	// Create a watch stream for either the database or collection.
+	stream, err := dataSource.Watch(ctx, mongo.Pipeline{})
+	if err != nil {
+		return fmt.Errorf("Error setting up changeStream: %s", err)
+	}
+	defer stream.Close(ctx)
+
+	// Watch and process changes.
+	a.processChanges(ctx, stream)
 	return nil
 }
+
+// Returns the appropriate dataSource: database or collection.
+func getDataSource(client *mongo.Client, db string, coll string) dataSource {
```

Comment:
make it a method part of the adapter? i.e., add (a *mongoDbAdapter)

Reply:
I added it to the code, since I think that's what you mentioned in the call?